File: ruby/merge_sort/merge_sort.rb
def merge_sort_recurse(ary, copy, left, right)
  size = right - left
  mid = (left + right) / 2
  return if (size == 1) # can't sort only 1 element

  # divide in half and sort the sections
  if (size > 2)
    merge_sort_recurse(ary, copy, left, mid)
    merge_sort_recurse(ary, copy, mid, right)
  end

  # after sorting our l/r sections
  # copy the items from the sections, in ascending order
  # into a temp ary
  lbegin = left
  rbegin = mid 
  for i in 0...size
    if (lbegin == mid) # left is done, copy right
      copy[left + i] = ary[rbegin]
      rbegin += 1
    elsif (rbegin == right) # right is done, copy left
      copy[left + i] = ary[lbegin]
      lbegin += 1
    else # copy smallest of l/r
      if ary[lbegin] < ary[rbegin]
        copy[left + i] = ary[lbegin]
        lbegin += 1
      else
        copy[left + i] = ary[rbegin]
        rbegin += 1
      end
    end
  end

  # copy the order of the temp array back into the original array
  for i in 0...size
    ary[left + i] = copy[left + i]
  end
end

def merge_sort(ary)
  aryLen = ary.size
  copy = [*0...aryLen]
  merge_sort_recurse(ary, copy, 0, aryLen)
end


randMin = 10
randMax = 100
aryLen = 10
ary = [*0...aryLen].map { |e| randMin + rand(randMax - randMin) }
puts ary.to_s
merge_sort(ary)
puts ary.to_s

File: javascript/merge_sort/merge_sort.js
var randMin = 0;
var randMax = 10;
var aryLen = 10;
var ary = new Array(aryLen)
  .fill(true)
  .map((e,i) => aryLen - i - 1); // backwards array
console.log(ary);

function mergeSort(ary) {
  var copy = new Array(ary.length);
  mergeSortRecurse(ary, copy, 0, ary.length);
}

function mergeSortRecurse(ary, copy, left, right) {
  var size = Math.floor(right - left); // MUST BE FLOORED
  if (size == 1) {
    return;
  }

  var mid = Math.floor((left + right) / 2); // MUST BE FLOORED
  if (size > 2) {
    mergeSortRecurse(ary, copy, left, mid);
    mergeSortRecurse(ary, copy, mid, right);
  }

  var lbegin = left;
  var rbegin = mid;
  var i = 0;
  while (lbegin < mid && rbegin < right) {
    if (ary[lbegin] < ary[rbegin]) {
      copy[left + i] = ary[lbegin];
      lbegin++;
    } else {
      copy[left + i] = ary[rbegin];
      rbegin++;
    }
    i++;
  }
  while (lbegin < mid) {
    copy[left + i] = ary[lbegin];
    lbegin++;
    i++;
  }
  while (rbegin < right) {
    copy[left + i] = ary[rbegin];
    rbegin++;
    i++;
  }

  for (i = 0; i < size; i++) {
    ary[left + i] = copy[left + i];
  }
}

mergeSort(ary);
console.log(ary);